### PR TITLE
fix: Race condition on deleteFrom #203

### PR DESCRIPTION
**Pull request by gbrd**
Signed-off-by: Gaël Bréard <gael.breard@orange.com>

### CODE REVIEW

**Comment by scullxbones** [comment]
Thanks @gbrd - should be able to get a good look at this on the weekend

**Comment by scullxbones** [comment]
Yep I think this should work for casbah.

Does RxM have the same race condition?  It looks like it might.  Do you want to get that as well?  I'd like to keep both drivers at parity.  If not, I can take a swing at it.

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala [hunk 1]
```diff
@@ -159,31 +160,38 @@ class RxMongoJournaller(driver: RxMongoDriver) extends MongoPersistenceJournalli
...
+      query = journalRangeQuery(persistenceId, toSequenceNr, toSequenceNr)
+      update = BSONDocument(
+        "$pull" -> BSONDocument(
+          EVENTS -> BSONDocument(
+            PROCESSOR_ID -> persistenceId,
+            SEQUENCE_NUMBER -> BSONDocument("$lte" -> toSequenceNr))),
+        "$set" -> BSONDocument(FROM -> (toSequenceNr + 1)))
+
+      wrUpdate <- journal.update(query, update, writeConcern, upsert = false, multi = true)
+      if wrUpdate.ok
```
Comment (scullxbones) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: I think this looks like a good translation except for the `DuplicateKey` race condition handling.

Comment (gbrd) [reply rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: It seems to be no catch for DuplicateKey in setMaxSequenceMetadata either. (#151 fix)
Thus I through it was different for rxmongo.
I will try to have a look

Comment (gbrd) [reply rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: Do you know what would be the error (and how to catch it) for a duplicatekey error with rxmongo driver ? (on GenericCollection.update() method)

Comment (scullxbones) [reply rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: I think there's an error code on the write result object that comes back.

According to `com.mongodb.ErrorCategory` in the java driver, the duplicate key codes are `11000, 11001, 12582`

**Comment by gbrd** [comment]
note:  if wrUpdate.ok is useless thanks to doc (and code), thus removed 

"If the write result actually indicates an error, the Future will be in a failed state (no need to check for WriteResult.ok)."
http://reactivemongo.org/releases/0.1x/documentation/tutorial/write-documents.html

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala (hunk 1 above)
Comment (gbrd) [reply rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: I added a recover to catch the duplicate key error.

**Comment by scullxbones** [comment]
> note: if wrUpdate.ok is useless thanks to doc (and code), thus removed

Good find!  Yep looks good to me.  Thanks for the PR!

**Pull request merged**